### PR TITLE
add more tests for disk-usage eviction

### PR DESCRIPTION
**Pull request by problame**

### CODE REVIEW

**Comment by problame** [comment]
@LizardWizzard this is ready for review.

File: libs/statvfs_ldpreload/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+            let avail_blocks = total_blocks - used_blocks;
+
+            // SAFETY: for the purposes of mocking, zeroed values for the fields which we
+            // don't set below are fine.
+            let mut ret = unsafe { MaybeUninit::<libc::statvfs64>::zeroed().assume_init() };
+            ret.f_bsize = blocksize;
+            ret.f_frsize = blocksize;
+            ret.f_blocks = total_blocks;
+            ret.f_bfree = avail_blocks;
```
Comment (LizardWizzard) [inline libs/statvfs_ldpreload/src/lib.rs:106]: Just to double check. Is it intended that both `f_bfree` and `f_bavail` are assigned with `avail_blocks`?

**Review by LizardWizzard — COMMENTED** [review]
Thanks for working on this!

I see the following downsides of the approach:

- The library is externally loaded only once, so its not possible (or quite tricky) to change the behavior while the test is running.
- Similar to previous one, if I get it right it is impossible to use this strategy in a unit test. Many unit tests are packed into one binary and library can be loaded only once.
- LD_PRELOAD feels hacky, some unsafe is needed to wire it up.
- How will it scale if we need to mock other apis?

See https://github.com/neondatabase/neon/pull/3809#issuecomment-1488681929

As I said it should be possible to have failpoint-like API for that, or even explicit dependency injection (more applicable in unit tests, needs to be combined with http api for integration ones).

Thoughts?

At the end of the day having tests is better than not having them.

**Comment by problame** [comment]
I think the alternative design for this is to make:
* pageserver provide a websocket to which test suite can connect
* if test suite is connected, pageserver doesn't to statvfs itself, but asks test suite for a statvfs over the websocket
* testsuite can provide whatever it wants

To me, the approach in this PR is less pain than figuring out websockets + rust webserver / async.

So, are you opposed to merging this in favor of some other testing design?

**Review by koivunej — COMMENTED** [review]
Paired with @problame on this so not approving.

**Comment by LizardWizzard** [comment]
> So, are you opposed to merging this in favor of some other testing design?

I'm not strictly opposed, but it feels suboptimal to me

**Comment by problame** [comment]
> Not sure why websocket is needed, why failpoint like approach wont work?

The problem is that we need the test suite to provide a mock statvfs response
that reflects the current disk-space utilization of the tenants directory.

But, the test suite doesn't know
* when the tenants directory changes (i.e., pageserver deletes layers)
* when the pageserver makes the statvfs call

The failpoints lib doesn't provide the ability for the test to tell whether
the pageserver has hit a given failpoint.

So, I can't see how the test suite can fill in a `result` failpoint with the
correct statvfs response.

---

I suppose, we could have two failpoints before the statvfs.

The first failpoint is a `sleep for X seconds`.
The second failpoint is a `return` failpoints that returns whatever mock response the test suite last punched in.

The test suite would start a background loop that punches in a mock response every X/2 seconds.

That guarantees that the mock response provided by the `return` failpoint always includes all pageserver actions that happened before the pageserver entered the `sleep for X seconds` failpoint.

Is that what you had in mind?

**Comment by LizardWizzard** [comment]
> The problem is that we need the test suite to provide a mock statvfs response
that reflects the current disk-space utilization of the tenants directory.

Is it required for that response to be accurate? I mean, if in the test I set statvfs call to return 100 when called first time and 80 when called second time (may not exactly match the failpoint api), will it work? We can adjust the value by looking at actual layers we have if they need to match. So something like that:

1. Generate some layers, assume we have two layers of size 100 and 50.
2. So first statvfs call should return something that will trigger `has_pressure` to return true. We know in advance how this "something" should look like. `avail_bytes` should be less than `config.min_avail_bytes` or `usage_pct` should be less than  `config.max_usage_pct`.
3. We expect that some layer should be evicted. It should be possible to determine which one will be evicted, we can calculate it from the layer map dump endpoint, right? So at the end of it we should know what is the size of the layer that needs to be evicted so we can made up second statfvs response in advance. So imagine we selected layer 50 to be least accessed one.
4. We should've evicted this layer, we can check it through the same dump api or on the fs
5. Check that layer was evicted and no longer exists on the filesystem.

It may be a bit tricky to get a test that is not flaky because background activity like gc or compaction can change access stat on the fly. This is another reason why unit test can be a good idea.

What do you think about this sequence? 

> Is that what you had in mind?

The sequence above is close to it, with the difference that I believe that we should be able to provide both returned values of statvfs in advance. 

I'll put it this way. As far as I understand we want to cut the corner here because we want to merge the original PR rather sooner. 

Citing myself from https://github.com/neondatabase/neon/pull/3809#issuecomment-1487374759:

> We can cut this corner now, but this shouldnt be the permanent state.

So having a test is a good thing. If you dont have time to polish the approach here lets merge it as is and create an issue for follow ups. I hope at some point we will get to testing an effect of spurious IO errors and similar stuff so having a more scalable approach would be needed.

Also there are some projects that build on FUSE and do most of the things we need:
- https://github.com/scylladb/charybdefs
- https://github.com/ligurio/unreliablefs (this one has good platform compatibility story)

**Comment by problame** [comment]
> Is it required for that response to be accurate? I mean, if in the test I set statvfs call to return 100 when called first time and 80 when called second time (may not exactly match the failpoint api), will it work?

1. It doesn't match the failpoint API semantics.
2. We _want_ to provide an accurate value here. Basically, the only reasons why we can't use statvfs directly are:
   2.1 The filesystem is shared among all concurrently running tests, and any other applications operating on it concurrently. We can't let the tests be affected by other processes that use the filesystem. Recipe for flakiness.
   2.2 There's no good way to make the test suite set up loop-mounts or the like, so that each test operates on its own isolated mountpoint.


> We expect that some layer should be evicted. It should be possible to determine which one will be evicted, we can calculate it from the layer map dump endpoint, right? So at the end of it we should know what is the size of the layer that needs to be evicted so we can made up second statfvs response in advance. So imagine we selected layer 50 to be least accessed one.

We currently avoid re-implementing the eviction policy in Python tests.
I'd like to keep it that way.

If you want to unit-test the eviction policy, that can be done as pure Rust unit test, without the need to mock statvfs at all, today.

> What do you think about this sequence?

I think it's more work and more brittle/flaky than the LD_PRELOAD.

**Comment by problame** [comment]
```
 =================================== FAILURES ===================================
_________________________ test_statvfs_pressure_usage __________________________
[gw14] linux -- Python 3.9.2 /github/home/.cache/pypoetry/virtualenvs/neon-_pxWMzVK-py3.9/bin/python
test_runner/regress/test_disk_usage_eviction.py:476: in test_statvfs_pressure_usage
    assert post_eviction_total_size <= 0.5 * total_size, "we requested max 50% usage"
E   AssertionError: we requested max 50% usage
E   assert 98680832 <= (0.5 * 193347584)
```

Hm, interesting. Can't repro this locally. Will debug tomorrow.

**Comment by LizardWizzard** [comment]
> 1. It doesn't match the failpoint API semantics.

Shouldnt be a big problem to have a mutex accessible through api with a vec of return values that are supplied when mock is called.

> We currently avoid re-implementing the eviction policy in Python tests.
I'd like to keep it that way.

Ok, we can avoid going into details and supply to mock two return values, first one that evaluates has_pressure to true and the second one which returns false. At the end we check that something was evicted. WDYT?

Looks like it shouldnt be flaky.

> If you want to unit-test the eviction policy, that can be done as pure Rust unit test, without the need to mock statvfs at all, today.

Do we have that listed somewhere?

**Comment by koivunej** [comment]
> So where we currently call statvfs, do something like this:

The LD_PRELOAD mock is used to test the actual background task. We could not get both before and after values with the suggested.

Looking at the time we've spent so far debugging the test failures in last commits, still unfinished, there is definitelty something interesting going on aside from the usual lifecycle issues. This would had been hidden should we had used a mock based with precalculated before and after.

**Comment by LizardWizzard** [comment]
> The LD_PRELOAD mock is used to test the actual background task. We could not get both before and after values with the suggested.

Why? See: https://github.com/neondatabase/neon/pull/3905#issuecomment-1489170737

**Comment by hlinnaka** [comment]
> > So where we currently call statvfs, do something like this:
> 
> The LD_PRELOAD mock is used to test the actual background task. We could not get both before and after values with the suggested.

I mean that the `mock_fstatfs(dir)` does exactly the same thing as the LD_PRELOADed fstatvfs in this patch. The only difference is that it's linked directly into the pageserver binary, instead of LD_PRELOADing it. That seems more straightforward than LD_PRELOADing in the CI scripts etc.

File: .github/workflows/build_and_test.yml [hunk 2]
```diff
@@ -260,6 +260,16 @@ jobs:
...
+          cdylibs=$(
+            ${cov_prefix} cargo metadata $CARGO_FEATURES --format-version=1 --no-deps |
+            jq -r '.packages[].targets[] | select(.kind | index("cdylib")) | .name'
+          )
+          for lib in $cdylibs; do
+            SRC=target/$BUILD_TYPE/lib$lib.so
+            DST=/tmp/neon/bin/lib$lib.so
+            cp "$SRC" "$DST"
+          done
+
```
[the remark this answers is not in the document]
Comment (problame) [reply .github/workflows/build_and_test.yml:272]: Needed to transport the libstatvfs_ldpreload.so into the next CI stage.

But, I guess this will be irrelevant as we'll switch to the proposed `mock_fstatfs` above.

File: test_runner/regress/test_disk_usage_eviction.py [hunk 3]
```diff
@@ -91,9 +95,38 @@ def warm_up_tenant(self, tenant_id: TenantId):
...
+
+        self.neon_env.pageserver.start(
+            overrides=(
+                "--pageserver-config-override=disk_usage_based_eviction="
+                + enc.dump_inline_table(disk_usage_config).replace("\n", " "),
+            ),
+        )
+
+        def statvfs_called():
+            assert self.neon_env.pageserver.log_contains(".*running mocked statvfs.*" + magic)
```
Comment (hlinnaka) [inline test_runner/regress/test_disk_usage_eviction.py:123]: This 'magic' feels like overkill, I think checking for the string "running mocked statvfs" is enough

File: pageserver/src/tenant/tasks.rs [hunk 4]
```diff
@@ -244,14 +244,12 @@ pub(crate) async fn random_init_delay(
 ) -> Result<(), Cancelled> {
     use rand::Rng;
 
+    if period == Duration::ZERO {
+        return Ok(());
+    }
+
```
Comment (hlinnaka) [inline pageserver/src/tenant/tasks.rs:250]: Maybe extract this to a separate PR. It seems worth drawing some extra attention to.

File: test_runner/regress/test_disk_usage_eviction.py [hunk 5]
```diff
@@ -365,3 +433,99 @@ def poor_mans_du(
...
+
+    env.pageserver_start_with_disk_usage_eviction(
+        period="1s",
+        max_usage_pct=33,
+        min_avail_bytes=0,
+        mock_behavior={
+            "type": "Success",
+            "blocksize": blocksize,
+            "total_blocks": total_blocks,
+            "name_filter": ".*__.*",
```
Comment (hlinnaka) [inline test_runner/regress/test_disk_usage_eviction.py:480]: What's point of this name_filter?

**Review by hlinnaka — COMMENTED** [review]
I appreciate having tests many error cases too, like with Broken tenants

File: pageserver/src/tenant/tasks.rs (hunk 4 above)
Comment (problame) [reply pageserver/src/tenant/tasks.rs:250]: https://github.com/neondatabase/neon/pull/3914

File: test_runner/regress/test_disk_usage_eviction.py (hunk 5 above)
Comment (problame) [reply test_runner/regress/test_disk_usage_eviction.py:496]: To just include layer files in the size calculation. Avoids having to account for metadata file sizes in the test code. Will add a comment.

Comment (hlinnaka) [reply test_runner/regress/test_disk_usage_eviction.py:488]: The metadata files are not very large,maybe just add some slop to the assertions in the tests to tolerate them?

File: pageserver/src/statvfs.rs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+impl Statvfs {
+    pub fn get(tenants_dir: &Path, mocked: Option<&mock::Behavior>) -> nix::Result<Self> {
+        if let Some(mocked) = mocked {
+            Ok(Statvfs::Mock(mock::get(tenants_dir, mocked)?))
+        } else {
+            Ok(Statvfs::Real(nix::sys::statvfs::statvfs(tenants_dir)?))
+        }
+    }
+
+    #[allow(clippy::useless_conversion)]
```
Comment (LizardWizzard) [inline pageserver/src/statvfs.rs:19]: Can you elaborate why these are needed?

File: test_runner/regress/test_disk_usage_eviction.py [hunk 7]
```diff
@@ -107,22 +134,18 @@ def eviction_env(request, neon_env_builder: NeonEnvBuilder, pg_bin: PgBin) -> It
...
-    pageserver_http.tenant_break(env.initial_tenant)
-    (broken_on_disk_before, _, _) = poor_mans_du(
-        env, timelines=[(broken_tenant_id, broken_timeline_id)]
-    )
-    env.pageserver.allowed_errors.append(
-        f".*extend_lru_candidates.*Tenant {broken_tenant_id} is not active. Current state: Broken"
-    )
-
-    timelines = []
+    # remove the initial tenant
```
Comment (LizardWizzard) [inline test_runner/regress/test_disk_usage_eviction.py:138]: Should it be possible to not create it in the first place? Maybe an argument to init_start, can be done separately.

**Review by LizardWizzard — APPROVED** [review]
Minor nits, LGTM otherwise. We can change/extend it when we need. Having automated tests for such a feature is nice

File: pageserver/src/statvfs.rs (hunk 6 above)
Comment (problame) [reply pageserver/src/statvfs.rs:19]: The `u64::try_from()` is required on macOS, where the block counts are u32 (c_ulong, actually).
On Linux x86_64, they're u64.

So, I put the `u64::try_from().unwrap()` in all 4 places for future-proofing other platforms.
I expect u64 to be sufficient for the foreseeable future.

Wanted to put a comment myself but didn't know where.
Suggestions?

File: test_runner/regress/test_disk_usage_eviction.py (hunk 7 above)
Comment (problame) [reply test_runner/regress/test_disk_usage_eviction.py:138]: It's not as straight-forward because `NeonEnv.tenant_id` is not `Optional[TenantId]`. Try to change it and see the plethora of mypy errors. And there are actually many tests that rely on an `initial_tenant` being there.

Here's a draft PR, will progress independently: https://github.com/neondatabase/neon/pull/3917

File: pageserver/src/statvfs.rs (hunk 6 above)
Comment (problame) [reply pageserver/src/statvfs.rs:24]: pushed a commit
